File: MetalSampleCodeLibrary/LightingTechniques/RenderingASceneWithDeferredLightingInObjectiveC/Renderer/AAPLRenderer_SinglePassDeferred.m
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
Implementation of the renderer class which performs Metal setup and per frame rendering for a
 single pass deferred renderer used for iOS & tvOS devices
*/

#import "AAPLRenderer_SinglePassDeferred.h"

// Include header shared between C code here, which executes Metal API commands, and .metal files
#import "AAPLShaderTypes.h"

@implementation AAPLRenderer_SinglePassDeferred
{
    id <MTLRenderPipelineState> _lightPipelineState;

    MTLRenderPassDescriptor *_viewRenderPassDescriptor;

    MTLStorageMode _GBufferStorageMode;

    MTKView *_view;
}

/// Perform single pass deferred renderer specific initialization
- (nonnull instancetype)initWithMetalKitView:(nonnull MTKView *)view
{
    self = [super initWithMetalKitView:view];

    if(self)
    {
        _view = view;

        if (@available( macOS 11, * ))
        {
            _GBufferStorageMode = MTLStorageModeMemoryless;
        }

        self.singlePassDeferred = YES;
        [self loadMetal];
        [self loadScene];
    }

    return self;
}

/// Create Metal render state objects specific to the single pass deferred renderer
- (void)loadMetal
{
    [super loadMetal];

    NSError *error;

    id <MTLLibrary> shaderLibrary = self.shaderLibrary;

    #pragma mark Point light render pipeline setup
    {
        MTLRenderPipelineDescriptor * renderPipelineDescriptor = [MTLRenderPipelineDescriptor new];
        renderPipelineDescriptor.colorAttachments[AAPLRenderTargetLighting].pixelFormat = self.view.colorPixelFormat;
        renderPipelineDescriptor.colorAttachments[AAPLRenderTargetAlbedo].pixelFormat = self.albedo_specular_GBufferFormat;
        renderPipelineDescriptor.colorAttachments[AAPLRenderTargetNormal].pixelFormat =  self.normal_shadow_GBufferFormat;
        renderPipelineDescriptor.colorAttachments[AAPLRenderTargetDepth].pixelFormat =  self.depth_GBufferFormat;
        renderPipelineDescriptor.depthAttachmentPixelFormat =  self.view.depthStencilPixelFormat;
        renderPipelineDescriptor.stencilAttachmentPixelFormat =  self.view.depthStencilPixelFormat;

        id <MTLFunction> lightVertexFunction = [shaderLibrary newFunctionWithName:@"deferred_point_lighting_vertex"];
        id <MTLFunction> lightFragmentFunction = [shaderLibrary newFunctionWithName:@"deferred_point_lighting_fragment_single_pass"];

        renderPipelineDescriptor.label = @"Light";
        renderPipelineDescriptor.vertexFunction = lightVertexFunction;
        renderPipelineDescriptor.fragmentFunction = lightFragmentFunction;
        _lightPipelineState = [self.device newRenderPipelineStateWithDescriptor:renderPipelineDescriptor
                                                                      error:&error];

        NSAssert(_lightPipelineState, @"Failed to create render pipeline state: %@", error);
    }

    #pragma mark GBuffer + View render pass descriptor setup
    _viewRenderPassDescriptor = [MTLRenderPassDescriptor new];
    _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetAlbedo].loadAction = MTLLoadActionDontCare;
    _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetAlbedo].storeAction = MTLStoreActionDontCare;
    _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetNormal].loadAction = MTLLoadActionDontCare;
    _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetNormal].storeAction = MTLStoreActionDontCare;
    _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetDepth].loadAction = MTLLoadActionDontCare;
    _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetDepth].storeAction = MTLStoreActionDontCare;
    _viewRenderPassDescriptor.depthAttachment.loadAction = MTLLoadActionClear;
    _viewRenderPassDescriptor.depthAttachment.storeAction = MTLStoreActionDontCare;
    _viewRenderPassDescriptor.stencilAttachment.loadAction = MTLLoadActionClear;
    _viewRenderPassDescriptor.stencilAttachment.storeAction = MTLStoreActionDontCare;
    _viewRenderPassDescriptor.depthAttachment.clearDepth = 1.0;
    _viewRenderPassDescriptor.stencilAttachment.clearStencil = 0;
}

/// MTKViewDelegate Callback: Respond to device orientation change or other view size change
- (void)mtkView:(nonnull MTKView *)view drawableSizeWillChange:(CGSize)size
{
    // The renderer base class allocates all GBuffers >except< lighting GBuffer (since with the
    // single-pass deferred renderer the lighting buffer is the same as the drawable)
    [super drawableSizeWillChange:size withGBufferStorageMode:_GBufferStorageMode];

    // Re-set GBuffer textures in the GBuffer render pass descriptor after they have been
    // reallocated by a resize
    _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetAlbedo].texture = self.albedo_specular_GBuffer;
    _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetNormal].texture = self.normal_shadow_GBuffer;
    _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetDepth].texture = self.depth_GBuffer;

    // Cannot set the depth stencil texture here since MTKView reallocates it *after* the
    // drawableSizeWillChange callback

    if(view.paused)
    {
        [view draw];
    }
}

/// Draw directional lighting, which, on with the single pass deferred renderer does not need
/// GBuffers set as textures as they do with the traditional deferred renderer
- (void)drawDirectionalLight:(nonnull id <MTLRenderCommandEncoder>)renderEncoder
{
    [renderEncoder pushDebugGroup:@"Draw Directional Light"];

    [super drawDirectionalLightCommon:renderEncoder];

    [renderEncoder popDebugGroup];
}

/// Setup single pass deferred renderer specific pipeline/  Then call common renderer code to apply
/// the point lights
- (void) drawPointLights:(id<MTLRenderCommandEncoder>)renderEncoder
{
    [renderEncoder pushDebugGroup:@"Draw Point Lights"];

    [renderEncoder setRenderPipelineState:_lightPipelineState];

    // Call common base class method after setting state in the renderEncoder specific to the
    // single-pass deferred renderer
    [super drawPointLightsCommon:renderEncoder];

    [renderEncoder popDebugGroup];
}

/// MTKViewDelegate callback: Called whenever the view needs to render
- (void)drawSceneToView:(nonnull MTKView *)view
{
    id<MTLCommandBuffer> commandBuffer = [self beginFrame];
    commandBuffer.label = @"Shadow commands";

    [super drawShadow:commandBuffer];

    // Commit commands so that Metal can begin working on non-drawable dependant work without
    // waiting for a drawable to become avaliable
    [commandBuffer commit];

    commandBuffer = [self beginDrawableCommands];
    commandBuffer.label = @"GBuffer & Lighting Commands";

    id<MTLTexture> drawableTexture = self.currentDrawableTexture;

    // The final pass can only render if a drawable is available, otherwise it needs to skip
    // rendering this frame.
    if(drawableTexture)
    {
        _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetLighting].texture = drawableTexture;
        _viewRenderPassDescriptor.depthAttachment.texture = self.view.depthStencilTexture;
        _viewRenderPassDescriptor.stencilAttachment.texture = self.view.depthStencilTexture;

        id <MTLRenderCommandEncoder> renderEncoder = [commandBuffer renderCommandEncoderWithDescriptor:_viewRenderPassDescriptor];
        renderEncoder.label = @"Combined GBuffer & Lighting Pass";

        [super drawGBuffer:renderEncoder];

        [self drawDirectionalLight:renderEncoder];

        [super drawPointLightMask:renderEncoder];

        [self drawPointLights:renderEncoder];

        [super drawSky:renderEncoder];

        [super drawFairies:renderEncoder];

        [renderEncoder endEncoding];
    }

    [self endFrame:commandBuffer];
}

#if SUPPORT_BUFFER_EXAMINATION
/// Enable (or disable) buffer examination mode
- (void)validateBufferExaminationMode
{
    // When in buffer examination mode, the renderer must allocate the GBuffers with
    // StorageModePrivate since the buffer examination manager needs the GBuffers written to main
    // memory to render them on screen later.
    // However, when a buffer examination mode is not enabled, the renderer only needs the GBuffers
    // in the GPU tile memory, so it can use StorageModeMemoryless to conserve memory.
    
    if(self.bufferExaminationManager.mode)
    {
        // Clear the background of the GBuffer when examining buffers.  When rendering normally
        // clearing is wasteful, but when examining the buffers, the backgrounds appear corrupt
        // making unclear what's actually rendered to the buffers
        _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetAlbedo].loadAction = MTLLoadActionClear;
        _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetNormal].loadAction = MTLLoadActionClear;
        _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetDepth].loadAction = MTLLoadActionClear;

        // Store results of all buffers to examine them.  This is wasteful when rendering
        // normally, but necessary to present them on screen.
        _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetAlbedo].storeAction = MTLStoreActionStore;
        _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetNormal].storeAction = MTLStoreActionStore;
        _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetDepth].storeAction = MTLStoreActionStore;
        _viewRenderPassDescriptor.depthAttachment.storeAction = MTLStoreActionStore;
        _viewRenderPassDescriptor.stencilAttachment.storeAction = MTLStoreActionStore;

        _GBufferStorageMode = MTLStorageModePrivate;
    }
    else
    {
        // When exiting buffer examination mode, return to efficient state settings
        _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetAlbedo].loadAction = MTLLoadActionDontCare;
        _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetNormal].loadAction = MTLLoadActionDontCare;
        _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetDepth].loadAction = MTLLoadActionDontCare;
        _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetAlbedo].storeAction = MTLStoreActionDontCare;
        _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetNormal].storeAction = MTLStoreActionDontCare;
        _viewRenderPassDescriptor.colorAttachments[AAPLRenderTargetDepth].storeAction = MTLStoreActionDontCare;
        _viewRenderPassDescriptor.depthAttachment.storeAction = MTLStoreActionDontCare;
        _viewRenderPassDescriptor.stencilAttachment.storeAction = MTLStoreActionDontCare;

        if (@available( macOS 11, *))
        {
            _GBufferStorageMode = MTLStorageModeMemoryless;
        }
    }

    // Force reallocation of GBuffers since storage mode will have changed.
    [self mtkView:_view drawableSizeWillChange:_view.drawableSize];
}

#endif

@end

File: MetalSampleCodeLibrary/LightingTechniques/RenderingASceneWithDeferredLightingInCplusplus/Renderer/AAPLRenderer_SinglePassDeferred.cpp
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
Implementation of the renderer class which performs Metal setup and per frame rendering for a
 single pass deferred renderer used for iOS & tvOS devices in addition to macOS devices with
 Apple Silicon.
*/

#include "AAPLRenderer_SinglePassDeferred.h"

// Include header shared between C code here, which executes Metal API commands, and .metal files
#include "Shaders/AAPLShaderTypes.h"

#include "AAPLUtilities.h"

Renderer_SinglePassDeferred::Renderer_SinglePassDeferred( MTL::Device* pDevice )
: Renderer( pDevice )
{
    m_singlePassDeferred = true;

    m_GBufferStorageMode = MTL::StorageModeMemoryless;

    loadMetalInternal();
    loadScene();
}

Renderer_SinglePassDeferred::~Renderer_SinglePassDeferred()
{
    m_pLightPipelineState->release();
    m_pViewRenderPassDescriptor->release();
}

void Renderer_SinglePassDeferred::loadMetalInternal()
{
    Renderer::loadMetal();
    NS::Error* pError = nullptr;

    #pragma mark Point light render pipeline setup
    {
        MTL::RenderPipelineDescriptor* pRenderPipelineDescriptor = MTL::RenderPipelineDescriptor::alloc()->init();

        pRenderPipelineDescriptor->setLabel( AAPLSTR( "Light" ) );
        pRenderPipelineDescriptor->colorAttachments()->object(RenderTargetLighting)->setPixelFormat(colorTargetPixelFormat());
        pRenderPipelineDescriptor->colorAttachments()->object(RenderTargetAlbedo)->setPixelFormat(m_albedo_specular_GBufferFormat);
        pRenderPipelineDescriptor->colorAttachments()->object(RenderTargetNormal)->setPixelFormat(m_normal_shadow_GBufferFormat);
        pRenderPipelineDescriptor->colorAttachments()->object(RenderTargetDepth)->setPixelFormat(m_depth_GBufferFormat);
        pRenderPipelineDescriptor->setDepthAttachmentPixelFormat(depthStencilTargetPixelFormat());
        pRenderPipelineDescriptor->setStencilAttachmentPixelFormat(depthStencilTargetPixelFormat());

        MTL::Library* pShaderLibrary = m_pDevice->newDefaultLibrary();

        AAPL_ASSERT( pShaderLibrary, "Failed to create default shader library" );

        MTL::Function* pLightVertexFunction = pShaderLibrary->newFunction( AAPLSTR( "deferred_point_lighting_vertex" ) );
        MTL::Function* pLightFragmentFunction = pShaderLibrary->newFunction( AAPLSTR( "deferred_point_lighting_fragment_single_pass" ) );

        AAPL_ASSERT( pLightVertexFunction, "Failed to load deferred_point_lighting_vertex" );
        AAPL_ASSERT( pLightFragmentFunction, "Failed to load deferred_point_lighting_fragment_single_pass" );

        pRenderPipelineDescriptor->setVertexFunction( pLightVertexFunction );
        pRenderPipelineDescriptor->setFragmentFunction( pLightFragmentFunction );

        m_pLightPipelineState = m_pDevice->newRenderPipelineState( pRenderPipelineDescriptor, &pError );

        AAPL_ASSERT_NULL_ERROR( pError, "Failed to create lighting render pipeline state" );
        
        pLightVertexFunction->release();
        pLightFragmentFunction->release();
        pRenderPipelineDescriptor->release();
        pShaderLibrary->release();
    }

    #pragma mark GBuffer + View render pass descriptor setup
    m_pViewRenderPassDescriptor = MTL::RenderPassDescriptor::alloc()->init();
    m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetAlbedo)->setLoadAction(MTL::LoadActionDontCare);
    m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetAlbedo)->setStoreAction(MTL::StoreActionDontCare);
    m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetNormal)->setLoadAction(MTL::LoadActionDontCare);
    m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetNormal)->setStoreAction(MTL::StoreActionDontCare);
    m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetDepth)->setLoadAction(MTL::LoadActionDontCare);
    m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetDepth)->setStoreAction(MTL::StoreActionDontCare);
    m_pViewRenderPassDescriptor->depthAttachment()->setLoadAction( MTL::LoadActionClear );
    m_pViewRenderPassDescriptor->depthAttachment()->setStoreAction( MTL::StoreActionDontCare) ;
    m_pViewRenderPassDescriptor->stencilAttachment()->setLoadAction( MTL::LoadActionClear );
    m_pViewRenderPassDescriptor->stencilAttachment()->setStoreAction( MTL::StoreActionDontCare );
    m_pViewRenderPassDescriptor->depthAttachment()->setClearDepth( 1.0 );
    m_pViewRenderPassDescriptor->stencilAttachment()->setClearStencil( 0 );

}

/// Respond to view size change
void Renderer_SinglePassDeferred::drawableSizeWillChange(const MTL::Size& size, MTL::StorageMode GBufferStorageMode)
{
    m_drawableSize = size;
    
    // The renderer base class allocates all GBuffers except lighting GBuffer, because with the
    // single-pass deferred renderer, the lighting buffer is the same as the drawable.
    Renderer::drawableSizeWillChange(size, m_GBufferStorageMode);

    // Re-set GBuffer textures in the GBuffer render pass descriptor after they have been
    // reallocated by a resize
    m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetAlbedo)->setTexture(m_albedo_specular_GBuffer);
    m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetNormal)->setTexture(m_normal_shadow_GBuffer);
    m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetDepth)->setTexture(m_depth_GBuffer);
}

/// Frame drawing routine
void Renderer_SinglePassDeferred::drawInView( bool isPaused, MTL::Drawable* pCurrentDrawable, MTL::Texture* pDepthStencilTexture )
{
    MTL::CommandBuffer* pCommandBuffer = beginFrame( isPaused );
    pCommandBuffer->setLabel( AAPLSTR( "Shadow commands" ) );

    drawShadow( pCommandBuffer );
    pCommandBuffer->commit();

    pCommandBuffer = beginDrawableCommands();
    pCommandBuffer->setLabel( AAPLSTR( "GBuffer & Lighting Commands" ) );

    MTL::Texture* pDrawableTexture = currentDrawableTexture( pCurrentDrawable );
    if ( pDrawableTexture )
    {
        m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetLighting)->setTexture( pDrawableTexture );
        m_pViewRenderPassDescriptor->depthAttachment()->setTexture( pDepthStencilTexture );
        m_pViewRenderPassDescriptor->stencilAttachment()->setTexture( pDepthStencilTexture );

        MTL::RenderCommandEncoder* pRenderEncoder = pCommandBuffer->renderCommandEncoder(m_pViewRenderPassDescriptor);
        pRenderEncoder->setLabel( AAPLSTR( "Combined GBuffer & Lighting Pass" ) );

        Renderer::drawGBuffer( pRenderEncoder );

        drawDirectionalLight( pRenderEncoder );

        Renderer::drawPointLightMask( pRenderEncoder );

        drawPointLights( pRenderEncoder );

        Renderer::drawSky( pRenderEncoder );

        Renderer::drawFairies( pRenderEncoder );

        pRenderEncoder->endEncoding();
    }

    endFrame( pCommandBuffer, pCurrentDrawable );
}

void Renderer_SinglePassDeferred::drawDirectionalLight(MTL::RenderCommandEncoder* pRenderEncoder)
{
    pRenderEncoder->pushDebugGroup( AAPLSTR( "Draw Directional Light" ) );

    Renderer::drawDirectionalLightCommon( pRenderEncoder );

    pRenderEncoder->popDebugGroup();
}

void Renderer_SinglePassDeferred::drawPointLights(MTL::RenderCommandEncoder* pRenderEncoder)
{
    pRenderEncoder->pushDebugGroup( AAPLSTR( "Draw Point Lights" ) );

    pRenderEncoder->setRenderPipelineState( m_pLightPipelineState );

    // Call the common base class method after setting the state in the renderEncoder specific to the
    // single-pass deferred renderer
    Renderer::drawPointLightsCommon( pRenderEncoder );

    pRenderEncoder->popDebugGroup();
}

#if SUPPORT_BUFFER_EXAMINATION

/// Set up render targets for display when buffer examination mode enabled. Set up target for
/// optimal rendering when buffer examination mode disabled.
void Renderer_SinglePassDeferred::validateBufferExaminationMode()
{
    // When in buffer examination mode, the renderer must allocate the GBuffers with
    // StorageModePrivate since the buffer examination manager needs the GBuffers written to main
    // memory to render them on screen later.
    // However, when a buffer examination mode is not enabled, the renderer only needs the GBuffers
    // in the GPU tile memory, so it can use StorageModeMemoryless to conserve memory.

    if( m_bufferExaminationManager->mode())
    {
        // Clear the background of the GBuffer when examining buffers. When rendering normally
        // clearing is wasteful, but when examining the buffers, the backgrounds appear corrupt
        // making unclear what's actually rendered to the buffers
        m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetAlbedo)->setLoadAction( MTL::LoadActionClear );
        m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetNormal)->setLoadAction( MTL::LoadActionClear );
        m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetDepth)->setLoadAction( MTL::LoadActionClear );

        // Store results of all buffers to examine them.  This is wasteful when rendering
        // normally, but necessary to present them on screen.
        m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetAlbedo)->setStoreAction( MTL::StoreActionStore );
        m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetNormal)->setStoreAction( MTL::StoreActionStore );
        m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetDepth)->setStoreAction( MTL::StoreActionStore );
        m_pViewRenderPassDescriptor->depthAttachment()->setStoreAction( MTL::StoreActionStore );
        m_pViewRenderPassDescriptor->stencilAttachment()->setStoreAction( MTL::StoreActionStore );

        m_GBufferStorageMode = MTL::StorageModePrivate;
    }
    else
    {
        // When exiting buffer examination mode, return to efficient state settings
        m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetAlbedo)->setLoadAction( MTL::LoadActionDontCare );
        m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetNormal)->setLoadAction( MTL::LoadActionDontCare );
        m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetDepth)->setLoadAction( MTL::LoadActionDontCare );
        m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetAlbedo)->setStoreAction( MTL::StoreActionDontCare );
        m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetNormal)->setStoreAction( MTL::StoreActionDontCare );
        m_pViewRenderPassDescriptor->colorAttachments()->object(RenderTargetDepth)->setStoreAction( MTL::StoreActionDontCare );
        m_pViewRenderPassDescriptor->depthAttachment()->setStoreAction( MTL::StoreActionDontCare );
        m_pViewRenderPassDescriptor->stencilAttachment()->setStoreAction( MTL::StoreActionDontCare );

        m_GBufferStorageMode = MTL::StorageModeMemoryless;
    }

    // Force reallocation of GBuffers.
    drawableSizeWillChange( m_drawableSize, m_GBufferStorageMode );
}

#endif // SUPPORT_BUFFER_EXAMINATION
